### PR TITLE
ISSUE-538: Added user defined CAN speed selection, by setting SJW, BR…

### LINKED ISSUE #538: Allow user defined PCAN speeds
Hello,

In pcan.py currently the PcanBus(BusABC) class allows to initialize the speeds which are defined in pcan_bitrate_objs.

I would like to extend the functionality to allow all possible speeds by setting SJW, BRP, TSEG2, TSEG1. So basically user defined speeds.

My idea is to create a new argument flag user_bit which either takes the old approach or creates out of the bitrate a TPCANHandle() with the above mentioned bits.

Best regards

### PR DESCRIPTION
**Pull request by ttcontrol-brunnthaler**
…P, TSEG2, TSEG1 with the bitrate argument

Closes #538.

### CODE REVIEW

**Comment by ttcontrol-FBU** [comment]
https://github.com/hardbyte/python-can/issues/538

File: can/interfaces/pcan/pcan.py [hunk 1]
```diff
@@ -232,10 +238,7 @@ def send(self, msg, timeout=None):
             msgType = PCAN_MESSAGE_STANDARD
 
         # create a TPCANMsg message structure
-        if platform.system() == 'Darwin':
```
Comment (felixdivo) [inline can/interfaces/pcan/pcan.py]: Why was this removed?

File: can/interfaces/pcan/pcan.py [hunk 2]
```diff
@@ -86,9 +86,15 @@ def __init__(self, channel='PCAN_USBBUS1', state=BusState.ACTIVE, bitrate=500000
             Bitrate of channel in bit/s.
             Default is 500 kbit/s.
 
+        :param bool user_bit:
+            If true use the user defined bitrate with SJW, BRP, TSEG2, TSEG1
+            Default is False
         """
         self.channel_info = channel
-        pcan_bitrate = pcan_bitrate_objs.get(bitrate, PCAN_BAUD_500K)
+        if not user_bit:
```
Comment (felixdivo) [inline can/interfaces/pcan/pcan.py]: Couldn't we solve this without another parameter as well?

```python
try:
    pcan_bitrate = pcan_bitrate_objs[bitrate]
except AttributeError:
    pcan_bitrate = TPCANHandle(bitrate)
```

File: can/interfaces/pcan/pcan.py [hunk 3]
```diff
@@ -86,9 +86,15 @@ def __init__(self, channel='PCAN_USBBUS1', state=BusState.ACTIVE, bitrate=500000
             Bitrate of channel in bit/s.
```
Comment (felixdivo) [inline can/interfaces/pcan/pcan.py]: We should document that this may not only be the bitrate of the channel, but also possibly the SJW, BRP, TSEG2 and TSEG1.

File: can/interfaces/pcan/pcan.py (hunk 2 above)
Comment (felixdivo) [reply can/interfaces/pcan/pcan.py]: This does remove that default parameter from the access to `pcan_bitrate_objs` but it's wasn't really helping anyways since the `__init__()` has already set it to a default value.

**Comment by felixdivo** [comment]
But thank you for the contribution! Where did you find the information about the custom bitrate?

File: can/interfaces/pcan/pcan.py (hunk 2 above)
Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: Yes your are right, I just thought about legacy issues if someone entered an incorrect baudrate, his program will fail (when he expected the 500k)

File: can/interfaces/pcan/pcan.py (hunk 1 above)
Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: Good question I didn't touch this one

Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: Good question I will add it

File: can/interfaces/pcan/pcan.py (hunk 2 above)
Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: Yes your are right, I just thought about legacy issues if someone entered an incorrect baudrate, his program will fail (when he expected the 500k)

**Comment by ttcontrol-FBU** [comment]
Hello felixdivo,
Sorry for answering this late!
There is this website: http://www.bittiming.can-wiki.info/ where you can find under the BOSCH 16 bit Bit Timing Register BTR how the custom CAN is built and where the bits are mapped too.
If you download the https://www.peak-system.com/Software-APIs.305.0.html PCANView you can enter the speed you want and get the Bus Timing Register as parameter back
Best regards

File: can/interfaces/pcan/pcan.py (hunk 2 above)
Comment (felixdivo) [reply can/interfaces/pcan/pcan.py]: Well, in the long run I guess we will catch more wrong usages than break old programs, so I would change it to the new behaviour.

**Comment by felixdivo** [comment]
Okay, nice to see that Wiki! It would probably be nice to link it somewhere.

File: can/interfaces/pcan/pcan.py (hunk 2 above)
Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: It should be KeyError right?

File: can/interfaces/pcan/pcan.py (hunk 3 above)
Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: I added it in the doc string

**Comment by ttcontrol-FBU** [comment]
@felixdivo sorry I overwrote some of the changes you made because the main develop branch was extended to CAN-FD could you redo the merge you did before.
Best regards

**Comment by felixdivo** [comment]
No sorry, only you can commit to your own branch. I think you should undo the rebase and apply b96a54f on top of a rebase from the GUI in GitLab or a new clean rebase done from CLI. Something seemed to have gone wrong.

**Comment by felixdivo** [comment]
And also by now I forgot what I changed. ^^

File: can/interfaces/pcan/pcan.py (hunk 2 above)
Comment (felixdivo) [reply can/interfaces/pcan/pcan.py]: Yep, sorry

File: can/interfaces/pcan/pcan.py (hunk 3 above)
Comment (felixdivo) [reply can/interfaces/pcan/pcan.py]: Nice!

File: can/interfaces/pcan/pcan.py [hunk 4]
```diff
@@ -86,9 +86,15 @@ def __init__(self, channel='PCAN_USBBUS1', state=BusState.ACTIVE, bitrate=500000
...
+        :param bool user_bit:
+            If true use the user defined bitrate with SJW, BRP, TSEG2, TSEG1
+            Default is False
         """
         self.channel_info = channel
-        pcan_bitrate = pcan_bitrate_objs.get(bitrate, PCAN_BAUD_500K)
+        if not user_bit:
+            pcan_bitrate = pcan_bitrate_objs.get(bitrate, PCAN_BAUD_500K)
+        else:
+            pcan_bitrate = TPCANHandle(bitrate)
```
[the remark this answers is not in the document]
Comment (ttcontrol-FBU) [reply can/interfaces/pcan/pcan.py]: Its just a type definition:
TPCANHandle               = c_ushort    # Represents a PCAN hardware channel handle
But you are right I should use the :
TPCANBaudrate          = c_ushort    # Represents a PCAN Baud rate register value

"(i.e. simple 500000 does not translate directly to the proper BTR0BTR1 value for 500KBaud)"
In this case it does because we look the default values (most used) up in a dictionary 
line 55: pcan_bitrate_objs
In the docstring I describe it quite well I think:
            This argument can also be used for a user defined CAN-speed (e.g.: 0x0012 for 1.33 MHz)
            For details on how to write a user speed register (SJW, TSEG 1/2, BRP) see:
            http://www.bittiming.can-wiki.info (NXP SJA1000)
Please tell me what I could add for better understandability

**Comment by ttcontrol-FBU** [comment]
@felixdivo
On 4th of april: @felixdivo Merge branch 'develop' into develop d8f3495
Could you redo this merge please?

**Comment by felixdivo** [comment]
I don't know how to do this, maybe @hardbyte has other rights? But maybe it's not possible since you applied changes on top of it.

File: CHANGELOG.txt [hunk 5]
```diff
@@ -1,21 +1,3 @@
-Version 3.2.0
```
Comment (hardbyte) [inline CHANGELOG.txt]: This change needs to be reversed

File: can/interfaces/pcan/pcan.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bhass1) [reply can/interfaces/pcan/pcan.py]: The problem I see is that `bitrate` is an integer representation of the bitrate (e.g. 500000). If user_bit is set, then line 97 becomes `pcan_bitrate = TPCANHandle(500000)` but in the dictionary, for 500K bitrate, `pcan_bitrate = TPCANBaudrate(0x1C) #or TPCANBaudrate(28)`. They both can't be right. 

The same issue comes up for the written example in the docstring. I'd expect `bitrate` to be 1330000 for 1.33M bitrate.

So the definition of what `bitrate` represents changes if `user_bit = True` or `user_bit = False`, and I don't think it's a good idea to do that.

Perhaps it would be better to define a new configuration type or calculate the register value of BTR0BTR1 within this method based on `bitrate`.

**Comment by bhass1** [comment]
Following up from @ttcontrol-FBU's comment above:
> If you download the https://www.peak-system.com/Software-APIs.305.0.html PCANView you can enter the speed you want and get the Bus Timing Register as parameter back

The tool name is actually "BaudTool"
image
It can calculate the BTR0/BTR1 registers based on a desired Baud Rate....
image

I'd suggest having a new configuration parameter specifically for PCAN (for now, perhaps this is more generally applicable to other peripherals) named `btr0btr1` that can be set to the hex value output by this "BaudTool" (e.g. 0x12).

So the user interface is either passed in as a dict key-value pair: `can.interface.Bus(btr0btr1=0x12)` or can be passed in from the config file:
```
#~/can.conf
[default]
interface = pcan
channel = PCAN_USBBUS1
btr0btr1 = 0x12
```

File: can/interfaces/pcan/pcan.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixdivo) [reply can/interfaces/pcan/pcan.py]: Is this solved? Once it is, I'd merge this PR.

[the remark this answers is not in the document]
Comment (bhass1) [reply can/interfaces/pcan/pcan.py]: Still discussing in Issue #538

**Comment by hardbyte** [comment]
@bhass1 are you happy to sign off on this? I'd like to hold off merging until I see your approving review.

I'm going to start getting develop ready for the `3.2` release so will target this feature for the next release -  either `3.3` or `4.0`.

**Comment by bhass1** [comment]
> @bhass1 are you happy to sign off on this? I'd like to hold off merging until I see your approving review.

I don't like the user-interface - it muddles the meaning of `bitrate`. I am too busy now to commit the changes to add a separate configuration parameter, `btr0btr1`. When is `3.3` due out?

**Comment by felixdivo** [comment]
The next release will now be 4.0. It will be some weeks until we release it, since there are quite a few major changes, like dropping support for all Python versions below 3.6 and reworking how exceptions are used. Take you time.

**Comment by marcel-kanter** [comment]
+1 for the separation of bitrate and btr0 btr1.
These settings seem to be available for many can controllers.

**Comment by felixdivo** [comment]
See #615

**Comment by felixdivo** [comment]
#615 is now merged.

**Comment by karlding** [comment]
Closing since this is superseeded by #625. Please reopen if necessary.
